lib/irq.c

```diff
@@ -0,0 +1,136 @@
+/*
+ * Copyright (c) 2019, Xilinx Inc. and Contributors. All rights reserved.
+ *
+ * SPDX-License-Identifier: BSD-3-Clause
+ */
+
+#include <errno.h>
+#include <metal/irq.h>
+#include <metal/list.h>
+#include <metal/utilities.h>
+#include <metal/atomic.h>
+
+/** Soft IRQs mapping it stores which IRQ triggers the soft IRQ */
+struct metal_irq_map {
+	int source_irq_base; /**< Source IRQ which triggers the soft IRQ */
+	int target_irq_base; /**< target soft IRQ */
+	int irq_num; /**< number of IRQs of the mapping */
+};
+
+/** Indicate which IRQ has been allocated */
+#define METAL_IRQ_MAX 2048
+static unsigned long
+metal_irq_allocated[metal_div_round_up(METAL_IRQ_MAX, METAL_BITS_PER_ULONG)];
+
+/** List of registered IRQ controller */
+static METAL_DECLARE_LIST(irq_cntrs);
+
+#define METAL_IRQ_MAX_MAPPINGS 1
+static struct metal_irq_map irq_maps[METAL_IRQ_MAX_MAPPINGS];
+
+int metal_irq_allocate(int irq_base, int irq_num)
+{
+	static atomic_flag flag;
+	int ret, i;
+
+	while(atomic_flag_test_and_set(&flag) != 0);
+	if (irq_base != METAL_IRQ_ANY) {
+		if (metal_bitmap_is_bit_set(metal_irq_allocated, irq_base)
+		    != 0) {
+			ret = -EINVAL;
+			goto err;
+		}
+		if ((irq_base + irq_num) >= METAL_IRQ_MAX) {
+			ret = -EINVAL;
+			goto err;
+		}
+	} else {
+		irq_base = metal_bitmap_next_clear_bit(metal_irq_allocated,
+						       0, METAL_IRQ_MAX);
+		if (irq_base == METAL_IRQ_MAX) {
+			ret = -EINVAL;
+			goto err;
+		}
+	}
+
+	for (i = irq_base; i < irq_base + irq_num; i++) {
+		metal_bitmap_set_bit(metal_irq_allocated, i);
+	}
+	ret = irq_base;
+
+err:
+	atomic_flag_clear(&flag);
+	return ret;
+}
+
+void metal_irq_register_controller(struct metal_irq_controller *cntr)
```

Comment:
I think we should remove the locking from this function and require that all interrupt controllers are registered before multiple threads are used.  Otherwise we would have to do locking in metal_irq_get_controller() as well.

Reply:
Agree.